In [1]:
# import the important library 
import numpy as np
import pandas as pd

In [2]:
# now read file with the help of Pandas library
data = pd.read_csv("C:\\Users\\ccpoint\\Downloads\\sample_demographic_data.csv")# store the file in the data varible
data # now print 

Unnamed: 0           State      Zip
0              1  ANDHRA PRADESH      NaN
1              2  ANDHRA PRADESH      NaN
2              3  ANDHRA PRADESH  500 762
3              4  ANDHRA PRADESH      NaN
4              5  ANDHRA PRADESH      NaN
...          ...             ...      ...
8550        8551  ANDHRA PRADESH      NaN
8551        8552  ANDHRA PRADESH      NaN
8552        8553  ANDHRA PRADESH      NaN
8553        8554  ANDHRA PRADESH      NaN
8554        8555  ANDHRA PRADESH      NaN

[8555 rows x 3 columns]

In [3]:
data.isna().sum() # check the nan/mising value in the data

Unnamed: 0       0
State         1559
Zip           4523
dtype: int64

In [4]:
# now handle the missing value in the data
# remove the row at least the one nan value 
data=data.dropna()


In [5]:
data.isna().sum() # again check the nan value in the data

Unnamed: 0    0
State         0
Zip           0
dtype: int64

above the result not any missing value in the data

In [6]:
data.Zip.describe()

count     3021
unique    1711
top           
freq       132
Name: Zip, dtype: object

In [7]:
data.Zip.unique()

array(['500 762', '500457', '505327', ..., '732203', '723151', '585403'],
      dtype=object)

 in above output some value is wrong as '500 762'  O/P  array([**'500 762'**, '500457', '505327', ..., '732203', '723151', '585403'],

In [8]:
#detect if give value in Zip value i float or not flot 
def is_int(x): # define a function called is_float
    try: # try convert into float
        int(x)
    except: # otherwise false 
        return False
    return True

In [9]:
data[~data["Zip"].apply(is_int)] # look at the value where it is not valid float value

Unnamed: 0           State      Zip
2              3  ANDHRA PRADESH  500 762
25            26  ANDHRA PRADESH  508 213
28            29  ANDHRA PRADESH  522 020
33            34  ANDHRA PRADESH  531 163
34            35  ANDHRA PRADESH  531 163
...          ...             ...      ...
6151        6152  ANDHRA PRADESH  515 411
6204        6205  ANDHRA PRADESH  500 033
6499        6500  ANDHRA PRADESH  522 414
6584        6585  ANDHRA PRADESH  500 053
7045        7046  ANDHRA PRADESH  500 029

[1302 rows x 3 columns]

In [10]:
# store the valid  In Zip float value in the new data
new_data=data[data["Zip"].apply(is_int)]
new_data

Unnamed: 0           State     Zip
7              8  ANDHRA PRADESH  500457
9             10  ANDHRA PRADESH  505327
11            12  ANDHRA PRADESH  500361
12            13  ANDHRA PRADESH  500482
15            16     MAHARASHTRA  400079
...          ...             ...     ...
8524        8525               T  501505
8525        8526               A  508374
8526        8527               T  504106
8527        8528               W       0
8528        8529               K  585403

[1719 rows x 3 columns]

In [11]:
new_data.isna().sum() # check any nan value in the data

Unnamed: 0    0
State         0
Zip           0
dtype: int64

In [12]:
#convert object datatype into int data type
new_data["Zip"] = new_data["Zip"].astype('int64')
new_data.dtypes


C:\Users\ccpoint\AppData\Local\Temp\ipykernel_1484\1082117428.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data["Zip"] = new_data["Zip"].astype('int64')


Unnamed: 0     int64
State         object
Zip            int64
dtype: object

In [13]:
new_data

Unnamed: 0           State     Zip
7              8  ANDHRA PRADESH  500457
9             10  ANDHRA PRADESH  505327
11            12  ANDHRA PRADESH  500361
12            13  ANDHRA PRADESH  500482
15            16     MAHARASHTRA  400079
...          ...             ...     ...
8524        8525               T  501505
8525        8526               A  508374
8526        8527               T  504106
8527        8528               W       0
8528        8529               K  585403

[1719 rows x 3 columns]

**in the above result the pincode value length less than 6 digit but in india pin code length is 6 digits so solve it**

In [14]:
lnn=new_data['Zip'][new_data['Zip']<999999]# now store the Zip value have length not more than  6 digit
lnn


7       500457
9       505327
11      500361
12      500482
15      400079
         ...  
8524    501505
8525    508374
8526    504106
8527         0
8528    585403
Name: Zip, Length: 1693, dtype: int64

**above result no any nan value and value in float so data is clean now data is useful**

In [24]:
new_data["Zip"]=lnn # store the Zip value
new_data

C:\Users\ccpoint\AppData\Local\Temp\ipykernel_1484\2964655311.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data["Zip"]=lnn # store the Zip value


Unnamed: 0           State     Zip
7              8  ANDHRA PRADESH  500457
9             10  ANDHRA PRADESH  505327
11            12  ANDHRA PRADESH  500361
12            13  ANDHRA PRADESH  500482
15            16     MAHARASHTRA  400079
...          ...             ...     ...
8522        8523               A  534416
8524        8525               T  501505
8525        8526               A  508374
8526        8527               T  504106
8528        8529               K  585403

[1535 rows x 3 columns]

but above resut have a some  nan value now solve it

In [16]:
new_data.isna().sum() # check how many nan value 

Unnamed: 0     0
State          0
Zip           26
dtype: int64

In [17]:
new_data=new_data.dropna() # drop row have a nan value
new_data.isna().sum() # again check the nan value


Unnamed: 0    0
State         0
Zip           0
dtype: int64

In [25]:
lnn1=new_data['Zip'][new_data['Zip']>=100000]# now store the Zip value have length not less than  6 digit
lnn1


7       500457
9       505327
11      500361
12      500482
15      400079
         ...  
8522    534416
8524    501505
8525    508374
8526    504106
8528    585403
Name: Zip, Length: 1535, dtype: int64

##  now our data is clean and useful 

In [26]:
new_data["Zip"]=lnn1 # store the Zip value
new_data

C:\Users\ccpoint\AppData\Local\Temp\ipykernel_1484\1243914548.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  new_data["Zip"]=lnn1 # store the Zip value


Unnamed: 0           State     Zip
7              8  ANDHRA PRADESH  500457
9             10  ANDHRA PRADESH  505327
11            12  ANDHRA PRADESH  500361
12            13  ANDHRA PRADESH  500482
15            16     MAHARASHTRA  400079
...          ...             ...     ...
8522        8523               A  534416
8524        8525               T  501505
8525        8526               A  508374
8526        8527               T  504106
8528        8529               K  585403

[1535 rows x 3 columns]

In [20]:
new_data.isna().sum() # check how many nan value 


Unnamed: 0      0
State           0
Zip           158
dtype: int64

In [21]:
new_data=new_data.dropna() # drop row have a nan value
new_data.isna().sum() # again check the nan value

Unnamed: 0    0
State         0
Zip           0
dtype: int64

In [27]:
new_data

Unnamed: 0           State     Zip
7              8  ANDHRA PRADESH  500457
9             10  ANDHRA PRADESH  505327
11            12  ANDHRA PRADESH  500361
12            13  ANDHRA PRADESH  500482
15            16     MAHARASHTRA  400079
...          ...             ...     ...
8522        8523               A  534416
8524        8525               T  501505
8525        8526               A  508374
8526        8527               T  504106
8528        8529               K  585403

[1535 rows x 3 columns]

In [28]:
# now save data in new file
new_data.to_csv("C:\\Users\\ccpoint\\Desktop\\new__data1.csv",index=False)